packages/block-library/src/embed/edit.js

```diff
@@ -21,98 +21,42 @@ import classnames from 'classnames';
-	return class extends Component {
-		constructor() {
-			super( ...arguments );
-			this.switchBackToURLInput = this.switchBackToURLInput.bind( this );
-			this.setUrl = this.setUrl.bind( this );
-			this.getMergedAttributes = this.getMergedAttributes.bind( this );
-			this.setMergedAttributes = this.setMergedAttributes.bind( this );
-			this.getResponsiveHelp = this.getResponsiveHelp.bind( this );
-			this.toggleResponsive = this.toggleResponsive.bind( this );
-			this.handleIncomingPreview = this.handleIncomingPreview.bind(
-				this
-			);
-
-			this.state = {
-				editingURL: false,
-				url: this.props.attributes.url,
-			};
-
-			if ( this.props.preview ) {
-				this.handleIncomingPreview();
-			}
-		}
-
-		handleIncomingPreview() {
-			this.setMergedAttributes();
-			if ( this.props.onReplace ) {
-				const upgradedBlock = createUpgradedEmbedBlock(
-					this.props,
-					this.getMergedAttributes()
-				);
-				if ( upgradedBlock ) {
-					this.props.onReplace( upgradedBlock );
-				}
-			}
-		}
-
-		componentDidUpdate( prevProps ) {
-			const hasPreview = undefined !== this.props.preview;
-			const hadPreview = undefined !== prevProps.preview;
-			const previewChanged =
-				prevProps.preview &&
-				this.props.preview &&
-				this.props.preview.html !== prevProps.preview.html;
-			const switchedPreview =
-				previewChanged || ( hasPreview && ! hadPreview );
-			const switchedURL =
-				this.props.attributes.url !== prevProps.attributes.url;
-
-			if ( switchedPreview || switchedURL ) {
-				if ( this.props.cannotEmbed ) {
-					// We either have a new preview or a new URL, but we can't embed it.
-					if ( ! this.props.fetching ) {
-						// If we're not fetching the preview, then we know it can't be embedded, so try
-						// removing any trailing slash, and resubmit.
-						this.resubmitWithoutTrailingSlash();
-					}
-					return;
-				}
-				this.handleIncomingPreview();
-			}
-		}
-
-		resubmitWithoutTrailingSlash() {
-			this.setState(
-				( prevState ) => ( {
-					url: prevState.url.replace( /\/$/, '' ),
-				} ),
-				this.setUrl
-			);
-		}
-
-		setUrl( event ) {
-			if ( event ) {
-				event.preventDefault();
-			}
-			const { url } = this.state;
-			const { setAttributes } = this.props;
-			this.setState( { editingURL: false } );
-			setAttributes( { url } );
-		}
-
-		/***
+	return function EmbedEditComponent( props ) {
+		const {
+			attributes,
+			cannotEmbed,
+			fetching,
+			isSelected,
+			onReplace,
+			preview,
+			setAttributes,
+			themeSupportsResponsive,
+			tryAgain,
+		} = props;
+
+		const [ url, setURL ] = useState( attributes.url );
+		const [ isEditingURL, setIsEditingURL ] = useState( false );
+
+		/**
 		 * @return {Object} Attributes derived from the preview, merged with the current attributes.
 		 */
-		getMergedAttributes() {
-			const { preview } = this.props;
-			const { className, allowResponsive } = this.props.attributes;
+		const getMergedAttributes = () => {
```

Comment:
These should probably be standard function declarations rather than anonymous functions, to ensure they have names in debug traces. Otherwise, I think they'll just appear as "anonymous" in error messages, which is annoying.

Reply:
Functions like this should actually display just fine in React dev tools. The JS engine will add the name as expected, so passing around `getMergedAttributes` will give you `getMergedAttributes.name === 'getMergedAttributes'`